src/os_common/os_fhandle.c

```diff
@@ -9,213 +9,103 @@
 
 /*
- * __fhandle_advise_notsup --
- *	POSIX fadvise unsupported.
- */
-static int
-__fhandle_advise_notsup(WT_SESSION_IMPL *session,
-    WT_FH *fh, wt_off_t offset, wt_off_t len, int advice)
-{
-	WT_UNUSED(session);
-	WT_UNUSED(fh);
-	WT_UNUSED(offset);
-	WT_UNUSED(len);
-	WT_UNUSED(advice);
-
-	/* Quietly fail, callers expect not-supported failures. */
-	return (ENOTSUP);
-}
-
-/*
  * __fhandle_allocate_notsup --
  *	POSIX fallocate unsupported.
  */
 static int
-__fhandle_allocate_notsup(
-    WT_SESSION_IMPL *session, WT_FH *fh, wt_off_t offset, wt_off_t len)
+__fhandle_allocate_notsup(WT_FILE_HANDLE *file_handle,
+    WT_SESSION *wt_session, wt_off_t offset, wt_off_t len)
 {
+	WT_SESSION_IMPL *session;
+
+	session = (WT_SESSION_IMPL *)wt_session;
 	WT_UNUSED(offset);
 	WT_UNUSED(len);
-	WT_RET_MSG(session, ENOTSUP, "%s: file-allocate", fh->name);
+	WT_RET_MSG(session, ENOTSUP, "%s: file-allocate", file_handle->name);
 }
 
 /*
- * __fhandle_close_notsup --
- *	ANSI C close/fclose unsupported.
+ * __fhandle_method_finalize --
+ *	Initialize any NULL WT_FH structure methods to not-supported. Doing
+ *	this means that custom file systems with incomplete implementations
+ *	won't dereference NULL pointers.
  */
 static int
-__fhandle_close_notsup(WT_SESSION_IMPL *session, WT_FH *fh)
+__fhandle_method_finalize(
+    WT_SESSION_IMPL *session, WT_FILE_HANDLE *handle, bool readonly)
 {
-	WT_RET_MSG(session, ENOTSUP, "%s: file-close", fh->name);
-}
-
-/*
- * __fhandle_lock_notsup --
- *	Lock/unlock a file unsupported.
- */
-static int
-__fhandle_lock_notsup(WT_SESSION_IMPL *session, WT_FH *fh, bool lock)
-{
-	WT_UNUSED(lock);
-	WT_RET_MSG(session, ENOTSUP, "%s: file-lock", fh->name);
-}
-
-/*
- * __fhandle_map_notsup --
- *	Map a file unsupported.
- */
-static int
-__fhandle_map_notsup(WT_SESSION_IMPL *session,
-    WT_FH *fh, void *p, size_t *lenp, void **mappingcookie)
-{
-	WT_UNUSED(p);
-	WT_UNUSED(lenp);
-	WT_UNUSED(mappingcookie);
-	WT_RET_MSG(session, ENOTSUP, "%s: file-map", fh->name);
-}
-
-/*
- * __fhandle_map_discard_notsup --
- *	Discard a section of a mapped region unsupported.
- */
-static int
-__fhandle_map_discard_notsup(
-    WT_SESSION_IMPL *session, WT_FH *fh, void *p, size_t len)
-{
-	WT_UNUSED(p);
-	WT_UNUSED(len);
-	WT_RET_MSG(session, ENOTSUP, "%s: file-map-discard", fh->name);
-}
+#define	WT_HANDLE_METHOD_REQ(name)					\
+	if (handle->name == NULL)					\
+		WT_RET_MSG(session, EINVAL,				\
+		    "a %s handle method must be configured", #name)
+
+	WT_HANDLE_METHOD_REQ(close);
+	/* fadvise is not required */
+	if (handle->fallocate == NULL)
+		handle->fallocate = __fhandle_allocate_notsup;
+	WT_HANDLE_METHOD_REQ(lock);
+	/* map is not required */
+	/* map_discard is not required */
+	/* map_preload is not required */
+	/* map_unmap is not required */
+	WT_HANDLE_METHOD_REQ(read);
+	WT_HANDLE_METHOD_REQ(size);
+	/* sync is not required */
+	if (!readonly) {
+		WT_HANDLE_METHOD_REQ(truncate);
+		WT_HANDLE_METHOD_REQ(write);
+	}
```

Comment:
I've taken a look at the "capability" question.  I don't mind using `NULL` for method handles if we can be sure we'll never try to indirect through them.  That said, can we classify all methods into one of the following categories?

* always required
* required to support `read_only=false`
* optional performance enhancement: may be used by WiredTiger if non-NULL (e.g., fallocate, mmap)

I think this would be simpler that having filesystem implementers have to think about what database features may be used (e.g., "required if logging is configured").

Are there methods that don't fall into one of the above categories?